### ISSUE TITLE
Issues with lazyloading

### ISSUE THREAD

**Issue by lionel-** [georgestagg/webR#15]
I'm having issues with macOS builds of webR with dev flang and the new emscripten-based builds. It appears these issues are caused by the way we currently work around lazyload databases.

From what I gather, the base package is loaded from R files rather than a lazyload database at some point, and this fails when defining `.leap.seconds` because the evaluation depends on values that are not loaded yet. This error happens early during startup when base is loaded via `R_ReplFile()`, and it is not reported to stdout or stderr. Instead, the problem manifests with various symbols not found, such as `R.version` or `.ArgsEnv`. I'm unsure why this happens on macOS and not within the docker environment.

I've explored a way of creating lazyload databases more systematically in https://github.com/lionel-/webR/compare/build-fixes...lionel-:lazyload. This is not perfect because it creates the database from the host package (itself lazyloaded) instead of from the omnibus loader file (except for the methods package which uses a different strategy because of the way it creates). Also it makes the makefiles patches more complicated. However, the build invocation on the webr side is correspondingly simpler, and it solves the build errors I encountered with the flang branch.

If in the longer term we manage to run R as a node.js application, then we would no longer need these cross-building workarounds. So I'm thinking the workarounds in the lazyload branch should be sufficient for now?

**Comment by georgestagg** [comment]
I've been experimenting with your branches and have seen the same issues. From what I can tell a large part of the problem was that the hack I set up in `007-emscripten-linklist-hack.diff` was not properly guarded by `#ifdef __EMSCRIPTEN__`. By luck this seemed to work OK in dockerised Linux but not in macOS - probably due to the different way virtual memory is setup and what values the pointers to the structures end up with.

I've fixed up the `linklist-hack` patch in [georgestagg/webR@flang-lazyload](https://github.com/georgestagg/webR/tree/flang-lazyload) along with pulling in your other commits. I'm still not quite getting it to build, but it seems closer (I now see `host/R-4.1.3/bin/R` being `Killed: 9` for some reason...) I'll try and look at things again later through the week.

I agree that the best way to deal with the lazyload issue would be to run webR as a node.js application and avoid the workarounds, but I don't yet know how easy or hard this will be. For the moment, your patches look good.

**Comment by georgestagg** [comment]
Unsure what I did differently (perhaps OOM before?) but I just tried again and flang-lazyload built OK for me with our combined commits. The R `kmeans` function works, which IIRC calls out to a Fortran `KMNS` function, so linking to and running fortran code seems to work in principle. Thanks again for all your effort so far!

I can reliably reproduce the `_FortranACharacterCompareScalar1` problem we've seen before using the R `glm` function, so I guess that's one of the next things to look at.

**Comment by lionel-** [comment]
> still have some missing symbols such as __fz_powi_1

yup I'm seeing these as well. They are unlikely to have been caused by the switch from R 4.1.2 to R 4.1.3 so it seems they appeared with the new emscripten-based build and config. I'll take a look into these missing symbols now.

In any case I'm unsure we can rely on flang for "production" builds yet. But it should at least be helpful for quicker development. For me the dragonegg version of emfc is very slow even on my intel mac.

> From what I can tell a large part of the problem was that the hack I set up in 007-emscripten-linklist-hack.diff was not properly guarded by #ifdef __EMSCRIPTEN__

oh hmm, locally I use a commit that removes this patch entirely. I was going to ask about this, is the patch still needed? webR seems to work fine without it, but I may have missed failing use cases. Do you have a reprex handy that the patch fixes?

**Comment by georgestagg** [comment]
> I'll take a look into these missing symbols now.

I wonder if we need some more intrinsics added to `llvmIntrinsics` in your forked flang. e.g. it has `atan` but not `atan2`. I guess we could also try to re-implement some of our own math functions to link to and fill in the gaps.

> In any case I'm unsure we can rely on flang for "production" builds yet.

I'm not sure either way to be honest, once we fill in the missing symbols I think we might be in a pretty good position to switch. My intuition feels like the dragonegg method is probably just as fragile.

>I was going to ask about this, is the patch still needed? 

Good question, it was one of my earlier patches. When I was first getting things up and running, R crashed while loading the `base` package without it. Since we've tweaked the build quite a bit since then I'll also rebuild now without the patch to test... Confirmed: it seems to work OK now, with both dragonegg and flang. That's brilliant news, great spot! I've reorganised the patches and removed that one in my latest commit to my copy of the `flang-lazyload` branch.

If we can deal with the missing symbols, I think I'd like to merge this all into main.
